src/librustdoc/doctest/make.rs

```diff
@@ -298,12 +303,14 @@ fn parse_source(
         match item.kind {
             ast::ItemKind::Fn(ref fn_item) if !info.has_main_fn => {
                 if item.ident.name == sym::main {
-                    info.has_main_fn = true;
+                    info.has_main_fn = is_top_level;
```

Comment:
```suggestion
                    info.has_main_fn = info.has_main_fn || is_top_level;
```

Also, please add this test case

````
/// ```
/// fn main() {
///     fn main() {}
/// }
/// ```
````

Reply:
Oof, good catch!